### PR TITLE
Target secret onchain

### PR DESCRIPTION
**Pull request by hackaugusto**
merge after: #2847

### CODE REVIEW

File: raiden/transfer/mediated_transfer/target.py [hunk 1]
```diff
@@ -39,12 +39,12 @@ def events_for_onchain_secretreveal(
...
         block_number,
     )
-    secret_known = channel.is_secret_known(
+    secret_known_offchain = channel.is_secret_known_offchain(
         channel_state.partner_state,
         transfer.lock.secrethash,
     )
 
-    if not safe_to_wait and secret_known:
+    if not safe_to_wait and secret_known_offchain:
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply raiden/transfer/mediated_transfer/target.py]: Answering my own question... when the partner knows the secret via the SecretReveal message but the transfer hasn't been unlocked, we keep checking on every block until it becomes unsafe to wait then we invoke registering the secret on-chain.
- In this code path, the transfer hasn't been unlocked as the current target task still exists.
- Also in this code path, the secret is known (see: target.handle_block).
So if what the above doing is:
-  Check if partner knows the secret. If so, and it's not safe to wait, then register the secret on-chain.
What should happen is:
- Check if partner knows the secret, it's not safe and wait and that the secret hasn't been already registered on chain... if so, register the secret on-chain.

[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/transfer/mediated_transfer/target.py]: > Check if partner knows the secret. If so, and it's not safe to wait, then register the secret on-chain. which on the next block, will happen again (please correct me if i am wrong here).

I believe there is a part of code that would prevent this.

At [events_for_onchain_secretreveal](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/secret_registry.py#L12) we got [this](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/secret_registry.py#L22) condition:

` if get_status(channel_state) in CHANNEL_STATES_PRIOR_TO_CLOSED + (CHANNEL_STATE_CLOSED, ):`

In here the `get_status(channel_state)` as you can see [here](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/channel.py#L980) will return a `CHANNEL_STATE_SETTLED`, which would stop it from sending on on-chain transaction again.

This is because it remembers the last sent transaction and checks its status.

File: raiden/tests/unit/transfer/mediated_transfer/test_targetstate.py [hunk 2]
```diff
@@ -228,6 +242,54 @@ def test_handle_secretreveal():
...
+    assert factories.UNIT_SECRETHASH in channel_state.partner_state.secrethashes_to_lockedlocks
+
+    offchain_secret_reveal_iteration = target.state_transition(
+        state,
+        ReceiveSecretReveal(secret, initiator),
+        channel_state,
+        pseudo_random_generator,
+        block_number,
+    )
+    assert factories.UNIT_SECRETHASH in channel_state.partner_state.secrethashes_to_unlockedlocks
```
Comment (ulope) [inline raiden/tests/unit/transfer/mediated_transfer/test_targetstate.py]: Maybe useless double testing, but wouldn't it make sense to also assert the secrethash is no longer in `secrethashes_to_lockedlocks` at this point?

File: raiden/transfer/mediated_transfer/target.py [hunk 3]
```diff
@@ -84,12 +84,16 @@ def handle_inittarget(
...
+        safe_to_wait, _ = is_safe_to_wait(
             transfer.lock.expiration,
             channel_state.reveal_timeout,
             block_number,
         )
 
+        # If there is not enough time to safely unlock the lock on-chain
+        # silently let the transfer expire. The target task must be created to
+        # handle the ReceiveLockExpired state change, which will clear the
+        # expired lock.
```
Comment (ulope) [inline raiden/transfer/mediated_transfer/target.py]: I don't get how this comment makes sense here. Where is the target task created it talks about?

Comment (LefterisJP) [reply raiden/transfer/mediated_transfer/target.py]: So perhaps the comment can be reworded a bit but what I believe it means is:

- The target task is created [here](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/mediated_transfer/target.py#L85) under the `is_valid` branch.

- Afterwards we check if it's safe to wait, and if it is we send the `SecretRequest`. If it's not we still [return](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/mediated_transfer/target.py#L118) the target task.

- So then the created target task can [handle the expired lock](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/mediated_transfer/target.py#L267) and clear it. 

So in conclusion what he is trying to say there is that if we did not return the target task we would never be able to handle the expired lock in the future.

File: raiden/tests/unit/transfer/mediated_transfer/test_targetstate.py (hunk 2 above)
Comment (LefterisJP) [reply raiden/tests/unit/transfer/mediated_transfer/test_targetstate.py]: Added it in a commit

File: raiden/transfer/mediated_transfer/target.py [hunk 4]
```diff
@@ -131,41 +124,28 @@ def handle_inittarget(
...
-            channel.register_onchain_secret(
-                channel_state=channel_state,
-                secret=state_change.secret,
-                secrethash=state_change.secrethash,
-                secret_reveal_block_number=state_change.block_number,
-            )
-        else:
-            assert False, 'Got unexpected StateChange'
+    if valid_secret:
+        channel.register_offchain_secret(
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/transfer/mediated_transfer/target.py]: added it

File: raiden/transfer/mediated_transfer/target.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/transfer/mediated_transfer/target.py]: So this still can't happen. I wrote a [test](https://github.com/raiden-network/raiden/pull/2846/commits/47af159e30d49052fb0215c5d2d9d3b54ec6bb10) for it.

The reason is that the [secret_known_offchain](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/mediated_transfer/target.py#L42) check fails [here](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/mediated_transfer/target.py#L42).

That is because when we [register the secret onchain](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/channel.py#L1574) we also [delete the lock](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/channel.py#L1549). This will remove it from the indexing structures and the next time around cause the `secret_known_offchain` check to fail.
